src/representation_theory/rep_hom.lean

```diff
@@ -0,0 +1,589 @@
+
+-- ext_ring
+end
+
+section
+
+variables
+  {V₃ : Type*} [add_comm_monoid V₃] [module k V₃]
+  {ρ₃ : representation k G V₃} (f₂ : ρ₂ →ᵣ ρ₃) (f : ρ →ᵣ ρ₂)
+
+/-- Composition of two rep_hom's is a rep_hom -/
+def comp : ρ →ᵣ ρ₃ :=
+{ to_fun := f₂ ∘ f,
+  map_smulG' := λ g x, by rw [comp_app, comp_app, map_smulG, map_smulG],
+  ..linear_map.comp f₂.to_linear_map f.to_linear_map }
+
+infixr ` ∘ᵣ `:80 := rep_hom.comp
+
+lemma comp_apply (x : V) : f₂.comp f x = f₂ (f x) := rfl
+
+@[simp, norm_cast] lemma coe_comp : (f₂.comp f : V → V₃) = f₂ ∘ f := rfl
+
+@[simp] theorem comp_id : f.comp id = f :=
+rep_hom.ext $ λ x, rfl
+
+@[simp] theorem id_comp : id.comp f = f :=
+rep_hom.ext $ λ x, rfl
+
+variables {f₂ f} {f₂' : ρ₂ →ᵣ ρ₃} {f' : ρ →ᵣ ρ₂}
+
+theorem cancel_right (hg : function.surjective f) :
+  f₂.comp f = f₂'.comp f ↔ f₂ = f₂' :=
+⟨λ h, ext $ hg.forall.2 (ext_iff.1 h), λ h, h ▸ rfl⟩
+
+theorem cancel_left (hf : function.injective f₂) :
+  f₂.comp f = f₂.comp f' ↔ f = f' :=
+⟨λ h, ext $ λ x, hf $ by rw [← comp_apply, h, comp_apply], λ h, h ▸ rfl⟩
+
+end
+
+/-- If a function `f'` is a left and right inverse of a rep_hom `f`, then `f'` is a rep_hom
+itself. -/
+def inverse
+  (f : ρ →ᵣ ρ₂) (f' : V₂ → V) (h₁ : left_inverse f' f) (h₂ : right_inverse f' f) :
+  ρ₂ →ᵣ ρ := by dsimp [left_inverse, function.right_inverse] at h₁ h₂; exact
+{ to_fun := f',
+  map_smulG' := λ g x, by {rw [←h₁ (ρ g (f' x)), map_smulG], simp [h₂]},
+  ..linear_map.inverse f.to_linear_map f' h₁ h₂ }
+
+end add_comm_monoid
+
+section add_comm_group
+
+variables
+  {k G V V₂ : Type*} [comm_semiring k] [monoid G]
+  [add_comm_group V] [module k V] [add_comm_group V₂] [module k V₂]
+  {ρ : representation k G V} {ρ₂ : representation k G V₂}
+  (f : ρ →ᵣ ρ₂)
+
+protected lemma map_neg (x : V) : f (- x) = - f x := map_neg f x
+
+protected lemma map_sub (x y : V) : f (x - y) = f x - f y := map_sub f x y
+
+-- compatible_smul
+
+end add_comm_group
+
+end rep_hom
+
+-- module
+-- distrib_mul_action_hom
+
+namespace is_rep_hom
+
+section add_comm_monoid
+
+variables
+  {k G V V₂ : Type*} [comm_semiring k] [monoid G]
+  [add_comm_monoid V] [module k V] [add_comm_monoid V₂] [module k V₂]
+  {ρ : representation k G V} {ρ₂ : representation k G V₂}
+
+/-- Convert an `is_rep_hom` predicate to a `rep_hom` -/
+def mk' (f : V → V₂) (H : is_rep_hom ρ ρ₂ f) : ρ →ᵣ ρ₂ :=
+{ to_fun := f, map_add' := H.1, map_smul' := H.2, map_smulG' := H.3 }
+
+@[simp] theorem mk'_apply {f : V → V₂} (H : is_rep_hom ρ ρ₂ f) (x : V) :
+  mk' f H x = f x := rfl
+
+lemma is_rep_hom_smul (c : k) :
+  is_rep_hom ρ ρ (λ (z : V), c • z) :=
+begin
+  refine is_rep_hom.mk (smul_add c) _ _,
+  { intros _ _,
+    rw [←mul_smul, mul_comm, mul_smul] },
+  { intros _ _,
+    rw [linear_map.map_smulₛₗ, ring_hom.id_apply] }
+end
+
+lemma is_rep_hom_smulG_one :
+  is_rep_hom ρ ρ (λ (z : V), ρ 1 z) :=
+begin
+  refine is_rep_hom.mk _ _ _;
+  { intros _ _,
+    simp only [map_one, linear_map.one_apply] }
+end
```

Comment:
That's a weird lemma since `λ (z : V), ρ 1 z` is just the identity map. Is there any use for it I'm missing?

Reply:
I'll removed this lemma